event.go

```diff
@@ -0,0 +1,225 @@
+package gomatrixserverlib
+
+import (
+	"encoding/json"
+	"fmt"
+	"golang.org/x/crypto/ed25519"
+	"time"
+)
+
+// An EventReference is a reference to a matrix event.
+type EventReference struct {
+	EventID     string
+	EventSHA256 Base64String
+}
+
+// An EventBuilder is used to build a new event.
+type EventBuilder struct {
+	// The user ID of the user sending the event.
+	Sender string `json:"sender"`
+	// The room ID of the room this event is in.
+	RoomID string `json:"room_id"`
+	// The type of the event.
+	EventType string `json:"type"`
+	// The state_key of the event if the event is a state event or nil if the event is not a state event.
+	EventStateKey *string `json:"state_key,omitempty"`
+	// The events that immediately preceeded this event in the room history.
+	PrevEvents []EventReference `json:"prev_events"`
+	// The events needed to authenticate this event.
+	AuthEvents []EventReference `json:"auth_events"`
+	// The event being redacted if this event is a "m.room.redaction".
+	Redacts  string `json:"redacts,omitempty"`
+	content  []byte
+	unsigned []byte
+}
+
+// SetContent sets the JSON content key of the event.
+func (eb *EventBuilder) SetContent(content interface{}) (err error) {
+	eb.content, err = json.Marshal(content)
+	return
+}
+
+// SetUnsigned sets the JSON unsigned key of the event.
+func (eb *EventBuilder) SetUnsigned(unsigned interface{}) (err error) {
+	eb.unsigned, err = json.Marshal(unsigned)
+	return
+}
+
+// An Event is a matrix event.
+// The event should always contain valid JSON.
+// If the event content hash is invalied then the event is redacted.
+// Redacted events contain only the fields covered by the event signature.
+type Event struct {
+	redacted  bool
+	eventJSON []byte
+}
+
+var emptyEventReferenceList = []EventReference{}
+
+// Build a new Event.
+// This is used when a local event is created on this server.
```

Comment:
A tiny bit of clarification would be good to explain **when** this should be called: notably after you've set all your bits and pieces on the `EventBuilder`. Is it safe to build the same event multiple times? We should say either way.